### PR TITLE
REGR: Series.nlargest with masked arrays

### LINKED ISSUE #42816: BUG: nlargest raises TypeError "No matching signature found" on Float64Dtype Series, versions >1.3.0
- [yes] I have checked that this issue has not already been reported.

- [yes] I have confirmed this bug exists on the latest version of pandas.

#### Code Sample, a copy-pastable example

```python
# this works:
pandas.Series(numpy.random.random(10)).nlargest(5)

# this works on pandas 1.2.5 but fails on 1.3.0 and 1.3.1.  All Float64DType Series seem to have the same issue
pandas.Series(numpy.random.random(10)).astype('Float64').nlargest(5)

```

#### Software/hardware
python version: 3.8.10
Ubuntu 20.04
(Intel Tiger Lake CPU)

#### Stack trace

```
TypeError                                 Traceback (most recent call last)
/tmp/ipykernel_213893/2592114079.py in <module>
----> 1 pandas.Series(numpy.random.random(10)).astype('Float64').nlargest(5)

~/.pyenv/versions/3.8.10/envs/myvenv/lib/python3.8/site-packages/pandas/core/series.py in nlargest(self, n, keep)
   3764         dtype: int64
   3765         """
-> 3766         return algorithms.SelectNSeries(self, n=n, keep=keep).nlargest()
   3767 
   3768     def nsmallest(self, n: int = 5, keep: str = "first") -> Series:

~/.pyenv/versions/3.8.10/envs/myvenv/lib/python3.8/site-packages/pandas/core/algorithms.py in nlargest(self)
   1217 
   1218     def nlargest(self):
-> 1219         return self.compute("nlargest")
   1220 
   1221     def nsmallest(self):

~/.pyenv/versions/3.8.10/envs/myvenv/lib/python3.8/site-packages/pandas/core/algorithms.py in compute(self, method)
   1285         # arr passed into kth_smallest must be contiguous. We copy
   1286         # here because kth_smallest will modify its input
-> 1287         kth_val = algos.kth_smallest(arr.copy(order="C"), n - 1)
   1288         (ns,) = np.nonzero(arr <= kth_val)
   1289         inds = ns[arr[ns].argsort(kind="mergesort")]

~/.pyenv/versions/3.8.10/envs/myvenv/lib/python3.8/site-packages/pandas/_libs/algos.pyx in pandas._libs.algos.__pyx_fused_cpdef()

TypeError: No matching signature found
```

Thanks!

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [x] closes #42816
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [x] whatsnew entry

### CODE REVIEW

File: doc/source/whatsnew/v1.3.2.rst [hunk 1]
```diff
@@ -21,7 +21,7 @@ Fixed regressions
 - Regression in :meth:`DataFrame.agg` when the ``func`` argument returned lists and ``axis=1`` (:issue:`42727`)
 - Fixed regression where :meth:`pandas.read_csv` raised a ``ValueError`` when parameters ``names`` and ``prefix`` were both set to None (:issue:`42387`)
 - Fixed regression in comparisons between :class:`Timestamp` object and ``datetime64`` objects outside the implementation bounds for nanosecond ``datetime64`` (:issue:`42794`)
--
+- Regression in :meth:`Series.nlargest` and :meth:`Series.nsmallest` with nullabe integer or float dtype (:issue:`41816`)
```
Comment (simonjayhawkins) [inline doc/source/whatsnew/v1.3.2.rst:24]:
```suggestion
- Regression in :meth:`Series.nlargest` and :meth:`Series.nsmallest` with nullable integer or float dtype (:issue:`41816`)
```

File: pandas/tests/series/methods/test_nlargest.py [hunk 2]
```diff
@@ -211,3 +211,14 @@ def test_nlargest_boolean(self, data, expected):
         result = ser.nlargest(1)
         expected = Series(expected)
         tm.assert_series_equal(result, expected)
+
+    def test_nlargest_nullable(self, any_nullable_numeric_dtype):
+        # GH#42816 NB so far we only handles cases without NAs
```
Comment (simonjayhawkins) [inline pandas/tests/series/methods/test_nlargest.py:216]: hmm, so this PR does not close the issue since this worked in 1.2.5?

Comment (jbrockmendel) [reply pandas/tests/series/methods/test_nlargest.py:216]: comment is out of date, this should now handle cases with NAs, though a test case needs to be added

File: pandas/core/algorithms.py [hunk 3]
```diff
@@ -1255,6 +1259,18 @@ def compute(self, method: str) -> Series:
 
         dropped = self.obj.dropna()
 
+        if is_extension_array_dtype(dropped.dtype):
```
Comment (jreback) [inline pandas/core/algorithms.py:1262]: why is this prefereable to hanlding on L1281 with the other dtypes?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/algorithms.py (hunk 3 above)
Comment (jbrockmendel) [reply pandas/core/algorithms.py:1262]: bc ensure_data does the wrong thing with MaskedArrays, np.asarray(obj) defaults to object dtype.

we _could_ kludge _ensure_data to work in cases where we dont have any pd.NAs, but doing it here lets us handle cases with NAs too.

Comment (jreback) [reply pandas/core/algorithms.py:1262]: > bc ensure_data does the wrong thing with MaskedArrays, np.asarray(obj) defaults to object dtype.

this is very unfortunate. I don't really like this approach here. i suppose its ok for a backport though this is na experimental type and so i don't consider this regression to be a big deal.

Comment (jbrockmendel) [reply pandas/core/algorithms.py:1262]: the more i work on it, the more i want to nuke NA from space

Comment (jreback) [reply pandas/core/algorithms.py:1262]: ok what is involved with changing this to a non-recursive formulation though? e.g. ensure_data *should* be able to handle NA (if not we will have other issues)

Comment (jbrockmendel) [reply pandas/core/algorithms.py:1262]: Three options:

1) have ensure_data special-case MaskedArray cases with no NAs, in which case they can just use the ndarray.  This fixes the regression (cases with NAs didnt use to work IIUC).  kluuuuudge
2) make algos.kth_smallest support object dtype (and not choke on pd.NA)
3) make the EA implement its own nlargest

I decided the approach here was less kludgy than those in part bc this function uses `obj.dropna()`, so the MaskedArray case actually is much simpler to implement than an arbitrary EA

Comment (jreback) [reply pandas/core/algorithms.py:1262]: why is 1 a kludge?

Comment (jbrockmendel) [reply pandas/core/algorithms.py:1262]: bc its special-casing for MaskedArray and special casing for `not values.isna().any()`

Comment (jbrockmendel) [reply pandas/core/algorithms.py:1262]: worse than that, it isnt `not values.isna().any()` but `not values._mask.any()`

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
note super thrilled with this, but does fix.

**Comment by jreback** [comment]
@meeseeksdev backport 1.3.x

File: doc/source/whatsnew/v1.3.2.rst [hunk 4]
```diff
@@ -23,6 +23,7 @@ Fixed regressions
 - Fixed regression where :meth:`pandas.read_csv` raised a ``ValueError`` when parameters ``names`` and ``prefix`` were both set to None (:issue:`42387`)
 - Fixed regression in comparisons between :class:`Timestamp` object and ``datetime64`` objects outside the implementation bounds for nanosecond ``datetime64`` (:issue:`42794`)
 - Fixed regression in :meth:`.Styler.highlight_min` and :meth:`.Styler.highlight_max` where ``pandas.NA`` was not successfully ignored (:issue:`42650`)
+- Regression in :meth:`Series.nlargest` and :meth:`Series.nsmallest` with nullable integer or float dtype (:issue:`41816`)
```
Comment (simonjayhawkins) [inline doc/source/whatsnew/v1.3.2.rst:26]: will need to change to 42816. will do that after backport is merged.

Comment (simonjayhawkins) [reply doc/source/whatsnew/v1.3.2.rst:26]: done in #42983

**Pull request merged**